fn/mvar.go

```diff
@@ -0,0 +1,205 @@
+package fn
+
+import (
+	"sync"
+	"sync/atomic"
+)
+
+// MVar[A any] is a structure that is designed to store a single value in an API
+// that dispenses with data races. Think of it as a box for a value.
+//
+// It has two states: full and empty.
+//
+// It supports two operations: take and put.
+//
+// The state transition rules are as follows:
+//  1. put while full blocks.
+//  2. put while empty sets.
+//  3. take while full resets.
+//  4. take while empty blocks.
+//  5. read while full nops.
+//  6. read while empty blocks.
+type MVar[A any] struct {
+	// value is the core state of the MVar
+	value *atomic.Pointer[A]
+
+	// mutex is a shared Mutex used by all of the Conds
+	mutex *sync.Mutex
+
+	// rCond is the Cond used to wake blocked Read operations.
+	rCond *sync.Cond
+
+	// tCond is the Cond used to wake blocked Take operations.
+	tCond *sync.Cond
+
+	// pCond is the Cond used to wake blocked Put operations.
+	pCond *sync.Cond
+
+	// reads is a WaitGroup that Put will wait for before signaling takers.
+	reads *sync.WaitGroup
+}
+
+// Zero initializes an MVar that has no values in it. In this state, TakeMVar
+// will block and PutMVar will immediately succeed.
+//
+// Zero : () -> MVar[A].
+func Zero[A any]() MVar[A] {
+	mutex := &sync.Mutex{}
+
+	return MVar[A]{
+		value: &atomic.Pointer[A]{},
+		mutex: mutex,
+		rCond: sync.NewCond(mutex),
+		tCond: sync.NewCond(mutex),
+		pCond: sync.NewCond(mutex),
+		reads: &sync.WaitGroup{},
+	}
+}
+
+// NewMVar initializes an MVar that has a value in it from the getgo. In this
+// state, TakeMVar will succeed immediately and PutMVar will block.
+//
+// NewMVar : A -> MVar[A].
+func NewMVar[A any](a A) MVar[A] {
+	z := Zero[A]()
+	z.value.Store(&a)
+	return z
+}
+
+// Take will wait for a value to be put into the MVar and then immediately
+// take it out.
+//
+// Take : MVar[A] -> A.
+func (m *MVar[A]) Take() A {
+	var val *A
+
+	// Here we use a Cond pattern and try to Swap nil into value. If we get
+	// a non-nil out it means we can actually finish the Take.
+	m.mutex.Lock()
+	for val = m.value.Swap(nil); val == nil; val = m.value.Swap(nil) {
+		m.tCond.Wait()
```

Comment:
Do we need to thread through a `context.Context` here to allow any blocked callers to automatically exit if a shutdown is initiated?

Reply:
Probably a good idea. We can make it optional and if they don't provide one then we just block til the heat death of the universe.

Reply:
Upon second look, even if we accept a context here it raises the question of whether we want to have an error value that can be returned from this function. It would impair the interface unless it was acceptable to panic, which I suspect it wouldn't be. Otherwise we now have an error type we need to add to the interface which would have to be checked all the time. If you think it's worth it to have it be cancellable like that but less ergonomic, then we can.

Reply:
Wouldn't we get a deadlock if there's no way to stop waiting?

Reply:
Isn't that true of a number of other synchronization tools too, including go channels? It is true that a misuse of this API will absolutely result in a deadlock. Suppose we do want to make these waits cancellable, though, are we willing to panic the thread, or do we really want to check for async cancellation everywhere? I see an argument for both.